piker/brokers/kraken.py

```diff
@@ -162,6 +222,129 @@ async def _public(
+        return resp
+
+    async def get_positions(
+        self,
+        data: Dict[str, Any] = {}
+    ) -> Dict[str, Any]:
+        data['ofs'] = 0
+        positions = {}
+        vols = {}
+        # Grab all trade history
+        while True:
+            resp = await self.kraken_endpoint('TradesHistory', data)
+            # grab the first 50 trades
+            if data['ofs'] == 0:
+                trades = resp['result']['trades']
+            # load the next 50 trades using dict constructor
+            # for speed
+            elif data['ofs'] == 50:
+                trades = dict(trades, **resp['result']['trades'])
+            # catch the end of the trades
+            elif resp['result']['trades'] == {}:
+                count = resp['result']['count']
+                break
+            # update existing dict if num trades exceeds 100
+            else: 
+                trades.update(resp['result']['trades'])
+            # increment the offset counter
+            data['ofs'] += 50
+            # To avoid exceeding API rate limit in case of a lot of trades
+            time.sleep(1)
+
+        # make sure you grabbed all the trades
+        assert count == len(trades.values())
+
+        # positions
+        ## TODO: Make sure to add option to include fees in positions calc
+        for trade in trades.values():
+            sign = -1 if trade['type'] == 'sell' else 1
+            try:
+                positions[trade['pair']] += sign * float(trade['cost'])
+                vols[trade['pair']] += sign * float(trade['vol'])
+            except KeyError:
+                positions[trade['pair']] = sign * float(trade['cost'])
+                vols[trade['pair']] = sign * float(trade['vol'])
+    
+        for pair in positions.keys():
+            asset_balance = vols[pair]
+            if asset_balance == 0:
+                positions[pair] = 0
+            else:
+                positions[pair] /= asset_balance
+
+        return positions, vols
+
+    async def submit_limit(
+        self,
+        oid: str,
+        symbol: str,
+        price: float,
+        action: str,
+        size: float,
+#        account: str,
+        reqid: int = None,
+    ) -> int:
+        """Place an order and return integer request id provided by client.
+
+        """
+        # Build order data for kraken api
+        data = {
+            "userref": reqid,
+            "ordertype": "limit",
+            "type": action,
+            "volume": str(size),
+            "pair": symbol,
+            "price": str(price),
+            # set to True test AddOrder call without a real submission
+            "validate": False
+        }
+        resp = await self.kraken_endpoint('AddOrder', data)
+        return resp
+
+    async def submit_cancel(
+        self,
+        reqid: str,
+    ) -> None:
+        """Send cancel request for order id ``reqid``.
+
+        """
+        # txid is a transaction id given by kraken
+        data = {"txid": reqid}
```

Comment:
you can probably just pass this inline instead of the static `dict` declaration.